p2p/src/peer.rs

```
@@ -0,0 +1,182 @@
+use std::collections::HashMap;
+use std::thread;
+
+use eyre::Result;
+use flume::{self, Receiver, Sender};
+
+use tendermint::node;
+use tendermint::public_key::PublicKey;
+
+use crate::message;
+use crate::transport::{Connection, Direction, StreamId};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+pub trait State: private::Sealed {}
+
+pub struct Connected<Conn> {
+    connection: Direction<Conn>,
+}
+impl<Conn> State for Connected<Conn> {}
+
+pub struct Running<Conn> {
+    connection: Direction<Conn>,
+    pub receiver: Receiver<message::Receive>,
+    senders: HashMap<StreamId, Sender<Box<dyn message::Outgoing>>>,
+}
+impl<Conn> State for Running<Conn> {}
+
+pub struct Stopped {
+    error: Option<Error>,
+}
+impl State for Stopped {}
+
+pub struct Peer<St>
+where
+    St: private::Sealed,
+{
+    pub id: node::Id,
+
+    pub state: St,
+}
+
+impl<Conn> From<Direction<Conn>> for Peer<Connected<Conn>>
+where
+    Conn: Connection,
+{
+    fn from(connection: Direction<Conn>) -> Peer<Connected<Conn>> {
+        let pk = match &connection {
+            Direction::Incoming(conn) => conn.public_key(),
+            Direction::Outgoing(conn) => conn.public_key(),
+        };
+
+        let id = match pk {
+            PublicKey::Ed25519(ed25519) => node::Id::from(ed25519),
+            _ => panic!(),
```

We should return an error and thus use `TryFrom` if the conversion can fail.